src/compas/datastructures/tree/tree.py

```diff
@@ -0,0 +1,373 @@
+    root : :class:`~compas.datastructures.TreeNode`
+        The root node of the tree.
+    nodes : list[:class:`~compas.datastructures.TreeNode`]
+        The nodes of the tree.
+    leaves : generator[:class:`~compas.datastructures.TreeNode`]
+        A generator of the leaves of the tree.
+
+    Examples
+    --------
+    >>> from compas.datastructures import Tree, TreeNode
+    >>> tree = Tree()
+    >>> root = TreeNode('root')
+    >>> branch = TreeNode('branch')
+    >>> leaf1 = TreeNode('leaf1')
+    >>> leaf2 = TreeNode('leaf2')
+    >>> tree.add(root)
+    >>> root.add(branch)
+    >>> branch.add(leaf1)
+    >>> branch.add(leaf2)
+    >>> print(tree)
+    <Tree with 4 nodes, 1 branches, and 2 leaves>
+    >>> tree.print()
+    <TreeNode root>
+        <TreeNode branch>
+            <TreeNode leaf2>
+            <TreeNode leaf1>
+
+    """
+
+    DATASCHEMA = {
+        "type": "object",
+        "properties": {
+            "name": {"type": "string"},
+            "root": TreeNode.DATASCHEMA,
+            "attributes": {"type": "object"},
+        },
+        "required": ["name", "root", "attributes"],
+    }
+
+    def __init__(self, name=None, attributes=None):
+        super(Tree, self).__init__()
+        self.name = name
+        self.attributes = attributes or {}
+        self._root = None
+
+    @property
+    def data(self):
+        return {
+            "name": self.name,
+            "root": self.root.data,
+            "attributes": self.attributes,
+        }
+
+    @classmethod
+    def from_data(cls, data):
+        tree = cls(data["name"], data["attributes"])
+        root = TreeNode.from_data(data["root"])
+        tree.add(root)
+        return tree
+
+    @property
+    def root(self):
+        return self._root
+
+    def add(self, node, parent=None):
+        """
+        Add a node to the tree.
+
+        Parameters
+        ----------
+        node : :class:`~compas.datastructures.TreeNode`
+            The node to add.
+        parent : :class:`~compas.datastructures.TreeNode`, optional
+            The parent node of the node to add.
+            Default is ``None``, in which case the node is added as a root node.
+
+        Returns
+        -------
+        None
+
+        Raises
+        ------
+        TypeError
+            If the node is not a :class:`~compas.datastructures.TreeNode` object.
+            If the supplied parent node is not a :class:`~compas.datastructures.TreeNode` object.
+        ValueError
+            If the node is already part of another tree.
+            If the supplied parent node is not part of this tree.
+            If the tree already has a root node, when trying to add a root node.
+
+        """
+        if not isinstance(node, TreeNode):
+            raise TypeError("The node is not a TreeNode object.")
+
+        if node.parent:
+            raise ValueError("The node is already part of another tree, remove it from that tree first.")
```

Comment:
can also be THIS tree, no?

Reply:
Sorry for the confusion, I adjusted a bit the sentence here, the idea is that if a node is already added somewhere (a parent link already exists, no matter it is this tree or another), we can ask user to explicitly remove that link first. Another option would be to not care about it and directly overwrite this link, which one you think is better?